### PR TITLE
fix(908): fix for the cmake tool chain removal function.

### PR DESCRIPTION
**Pull request by alirana01**
The calculated ID in the removal function is different than the one in the add

[Original reported issue](https://github.com/eclipse-cdt/cdt/issues/908)

### CODE REVIEW

**Comment by jonahgraham** [comment]
@betamaxbandit Can you review this please. It looks correct to me, but I think this is in your area.

File: cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/internal/CMakeToolChainManager.java [hunk 1]
```diff
@@ -156,20 +156,18 @@ public void addToolChainFile(ICMakeToolChainFile file) {
...
-				file.getProperty(CMakeBuildConfiguration.TOOLCHAIN_ID));
-		filesByToolChain.remove(tcId);
-
-		String n = ((CMakeToolChainFile) file).n;
-		if (n != null) {
-			Preferences prefs = getPreferences();
-			Preferences tcNode = prefs.node(n);
-			try {
+		try {
+			String tcId = makeToolChainId(file.getToolChain());
```
Comment (betamaxbandit) [inline cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/internal/CMakeToolChainManager.java:160]: Yes, this looks like a definite improvement.

I couldn't figure out why the original code was written the way it was.

I think in the past I've worked around this by assigning the TOOLCHAIN_TYPE and TOOLCHAIN_ID properties in my code that adds the toolchain file, just after calling org.eclipse.cdt.cmake.core.ICMakeToolChainManager.newToolChainFile(Path), like this:

			// toolchainFile.setProperty(ICBuildConfiguration.TOOLCHAIN_TYPE, tc.getTypeId());
			// toolchainFile.setProperty(ICBuildConfiguration.TOOLCHAIN_ID, tc.getId());

But this new code improves that situation so it's no longer necessary.

File: cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/internal/CMakeToolChainManager.java [hunk 2]
```diff
@@ -156,20 +156,18 @@ public void addToolChainFile(ICMakeToolChainFile file) {
...
+			String n = ((CMakeToolChainFile) file).n;
+			if (n != null) {
+				Preferences prefs = getPreferences();
+				Preferences tcNode = prefs.node(n);
 				tcNode.removeNode();
 				prefs.flush();
-			} catch (BackingStoreException e) {
-				Activator.log(e);
 			}
+		} catch (Exception e) {
```
Comment (betamaxbandit) [inline cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/internal/CMakeToolChainManager.java:169]: I would consider handling the CoreException, from org.eclipse.cdt.cmake.core.ICMakeToolChainFile.getToolChain(), separately. I think it's generally considered bad practice to catch the general Exception.

**Review by betamaxbandit — CHANGES_REQUESTED** [review]
LGTM.
Just an outstanding issue with the exception handling.

File: cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/internal/CMakeToolChainManager.java (hunk 2 above)
Comment (alirana01) [reply cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/internal/CMakeToolChainManager.java]: Thanks for review @betamaxbandit updated the exception block for CoreException and BackingStoreException separately.

File: cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/internal/CMakeToolChainManager.java (hunk 1 above)
Comment (jonahgraham) [reply cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/internal/CMakeToolChainManager.java:160]: Thanks @betamaxbandit for having a look. Marking this thread as resolved.

File: cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/internal/CMakeToolChainManager.java [hunk 3]
```diff
@@ -156,20 +156,20 @@ public void addToolChainFile(ICMakeToolChainFile file) {
...
+				Preferences tcNode = prefs.node(n);
 				tcNode.removeNode();
 				prefs.flush();
-			} catch (BackingStoreException e) {
-				Activator.log(e);
 			}
+		} catch (CoreException ce) {
+			Activator.log(ce);
+		} catch (BackingStoreException be) {
+			Activator.log(be);
```
Comment (jonahgraham) [inline cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/internal/CMakeToolChainManager.java:172]:
```suggestion
		} catch (CoreException | BackingStoreException e) {
			Activator.log(e);
```

**Review by jonahgraham — APPROVED** [review]
This LGTM. I am going to apply my minor suggested change and merge once CI approves too.

**Pull request merged**